### ISSUE TITLE
Loss spike during training phase

### ISSUE THREAD

**Issue by yzxyzh** [artidoro/qlora#84]
Hi!

I'm trying to experiment you lora on different datasets. I observe that the loss is generally decreasing for each test I am performing, but also contain strange spikes. The loss value and decrease pattern is very similar with full-parameter fine-tune, thus I assume the final result should be fine, but those spikes are unexplainable to me. Similar spike has never been observed during full parameter fine-tuning. Any idea what has caused these spikes?

the spikes are marked with red boxes.
<img width="392" alt="Screen Shot 2023-05-30 at 8 14 03 AM" src="https://github.com/artidoro/qlora/assets/6187703/b717d95f-6abe-406b-a566-6a679b377745">

Thanks

**Comment by baibaiw5** [comment]
> Hi! I meet the same problem

![image](https://github.com/artidoro/qlora/assets/233871/2d5f33f8-3928-400d-b8b5-c2bf690f6a21)

**Comment by Maxwell-Lyu** [comment]
Same here, but the trained model seems fine? (sort of)

**Comment by eggqq007** [comment]
because the qloar.py consume group_by_length.......

**Comment by artidoro** [comment]
Yes, group_by_length batches examples with similar length together and creates the oscillating pattern in the loss that you observe. The batches with shorter examples and those with longer ones have different loss values as they tend to correlate with "difficulty".

Using this setting is not necessary but will improve finetuning efficiency by reducing padding.
